config/reader.go

```diff
@@ -275,7 +282,7 @@ func readFromCloud(
 			if !errors.As(err, &context.DeadlineExceeded) {
 				return nil, err
 			}
-			if tls.certificate == "" || tls.privateKey == "" {
+			if errors.Is(err, errNoTLSCertificate) || errors.Is(err, errNoTLSPrivateKey) {
 				return nil, errors.Wrap(err, "error getting certificate data from cloud; try again later")
 			}
```

Comment:
This was a bug that I guess people never hit. `readCertificateDataFromCloudGRPC` only error when these are empty if `!signalingInsecure`, however, this errors always if there is any error in the `read` when it should warn in that case

Reply:
An alternative would be to add a `!signalingInsecure` to this if, but that feels... dangerous and brittle

Reply:
> I think the logic we want here should be is if there is no cert and signaling is not insecure, right?

I believe that the intended logic is actually: if this function returns one of those two error types, then we should exit, otherwise we can continue with an log warning.

The way that we used to detect if the function had that error type is by duplicating the same condition from inside readCertificateDataFromCloudGRPC (i.e., they both had the same empty-string comparisons).

However, the two conditions had gotten out of sync so there was an possibility that the readCertificateDataFromCloudGRPC would return a different error while signalingInsecure which would trigger this. 

The correct thing to do (I think) is to check the errors themselves. 

But this is your area of the codebase, so happy to revert to adding another condition to the if

Reply:
I have swapped to your proposal here: 5d49f037e

Either way LGTM. Happy with the explicit error check or the !signalingInsecure check

Reply:
@cheukt can you take a gander at 6810a8027

It reflects what we talked about offline and I think is better & easier to read

Reply:
I also moved up the check to be higher. There is no reason to be making requests to get tls info in insecure signaling.

Reply:
Running a manual test against prod and local right now.

Reply:
Manual test on prod and local app LGTM.

I tested with & without a cached config on both. Both behaved as expected